ibtk/src/math/PETScMatUtilities.cpp

```diff
@@ -806,10 +623,827 @@ void PETScMatUtilities::constructPatchLevelSCInterpOp(Mat& mat,
+
+                for (Box<NDIM>::Iterator b(box_to_iterate); b; b++)
+                {
+                    const CellIndex<NDIM>& i = b();
+                    if (patch_touches_cf_bdry && check_upper_dofs[axis] &&
+                        !is_cf_bdry_idx(i, upper_side_cf_bdry_box[axis]))
+                        continue;
+
+                    const SideIndex<NDIM> i_s(i, axis, SideIndex<NDIM>::Lower);
+                    const int dof_idx = (*dof_data)(i_s);
+#if !defined(NDEBUG)
+                    TBOX_ASSERT(dof_idx >= dof_lower && dof_idx < dof_upper);
+#endif
+                    box_local_dofs.push_back(dof_idx);
+                }
+            }
+            const int n_idx = static_cast<int>(box_local_dofs.size());
+            ISCreateGeneral(PETSC_COMM_SELF, n_idx, &box_local_dofs[0], PETSC_COPY_VALUES,
+                            &is_nonoverlap[0][is_counter]);
+
+            if (!there_is_overlap)
+            {
+                PetscObjectReference((PetscObject)is_nonoverlap[0][is_counter]);
+                is_overlap[0][is_counter] = is_nonoverlap[0][is_counter];
+            }
+            else
+            {
+                const Box<NDIM>& box_overlap = overlap_boxes[patch_counter][i];
+                Box<NDIM> side_box_overlap[NDIM];
+                int box_overlap_dofs_size = 0;
+                for (int axis = 0; axis < NDIM; ++axis)
+                {
+                    side_box_overlap[axis] = SideGeometry<NDIM>::toSideBox(box_overlap, axis);
+                    box_overlap_dofs_size += side_box_overlap[axis].size();
+                }
+                std::vector<int> box_overlap_dofs;
+                box_overlap_dofs.reserve(box_overlap_dofs_size);
+
+                for (int axis = 0; axis < NDIM; ++axis)
+                {
+                    for (Box<NDIM>::Iterator b(side_box_overlap[axis]); b; b++)
+                    {
+                        const CellIndex<NDIM>& i = b();
+                        const SideIndex<NDIM> i_s(i, axis, SideIndex<NDIM>::Lower);
+                        const int dof_idx = (*dof_data)(i_s);
+
+                        // We keep only those DOFs that are inside the
+                        // physical domain and on physical and c-f boundaries.
+                        // Some of the DOFs will be on other processors.
+                        if (dof_idx >= 0)
+                        {
+                            box_overlap_dofs.push_back(dof_idx);
+                        }
+                    }
+                }
+                const int n_idx = static_cast<int>(box_overlap_dofs.size());
+                ISCreateGeneral(PETSC_COMM_SELF, n_idx, &box_overlap_dofs[0], PETSC_COPY_VALUES,
+                                &is_overlap[0][is_counter]);
+            }
+            ++is_counter;
+        }
+    }
+
+#if !defined(NDEBUG)
+    TBOX_ASSERT(is_counter == n_subdomains);
+#endif
+
+    return;
+} // constructPatchLevelASMSubdomains_side
+
```

Comment:
(1) Regarding MSM, I don't think there will be any extra work, although perhaps that code will be more permissive about whether it is necessary to have nonoverlapping interior partitionings within local patches.

(2) Here is a test: If you make the subdomain size = patch size and an overlap of one cell, then I think that you should be able to replicate the box smoother FAC operator; performance should be nearly identical if you use the same subdomain solvers in both FAC implementations. If you make the subdomain size = 1 grid cell, then you should get something that is equivalent to using Jacobi as the smoother (which means that you probably should use damping).